packages/embeds/embed-core/src/embed.ts

```diff
@@ -310,58 +459,73 @@ export class Cal {
     if (attributes?.id) {
       existingEl = document.getElementById(attributes.id);
     }
-    let el: HTMLElement;
+    let el: FloatingButton;
     if (!existingEl) {
-      el = document.createElement("cal-floating-button");
+      el = document.createElement("cal-floating-button") as FloatingButton;
+      // It makes it a target element that opens up embed modal on click
       el.dataset.calLink = calLink;
-      el.dataset.calNamespace = this.namespace;
+      el.dataset.calNamespace = this.cal.namespace;
+      el.dataset.calOrigin = calOrigin ?? "";
       if (attributes?.id) {
         el.id = attributes.id;
       }
 
       document.body.appendChild(el);
     } else {
-      el = existingEl;
-    }
-
-    if (buttonText) {
-      el.setAttribute("data-button-text", buttonText);
+      el = existingEl as FloatingButton;
     }
-    el.setAttribute("data-hide-button-icon", "" + hideButtonIcon);
-    el.setAttribute("data-button-position", "" + buttonPosition);
-    el.setAttribute("data-button-color", "" + buttonColor);
-    el.setAttribute("data-button-text-color", "" + buttonTextColor);
+    const dataset = el.dataset;
+    dataset["buttonText"] = buttonText;
+    dataset["hideButtonIcon"] = "" + hideButtonIcon;
+    dataset["buttonPosition"] = "" + buttonPosition;
+    dataset["buttonColor"] = "" + buttonColor;
+    dataset["buttonTextColor"] = "" + buttonTextColor;
   }
 
-  modal({ calLink, config = {}, uid }: { calLink: string; config?: Record<string, string>; uid: number }) {
+  modal({
+    calLink,
+    calOrigin,
+    config = {},
+    uid,
+  }: {
+    calLink: string;
+    config?: PrefillAndIframeAttrsConfig;
+    uid?: string | number;
+    calOrigin?: string;
+  }) {
+    uid = uid || 0;
+
     const existingModalEl = document.querySelector(`cal-modal-box[uid="${uid}"]`);
     if (existingModalEl) {
       existingModalEl.setAttribute("state", "started");
       return;
     }
+    if (typeof config.iframeAttrs === "string" || config.iframeAttrs instanceof Array) {
+      throw new Error("iframeAttrs should be an object");
+    }
     config.embedType = "modal";
-    const iframe = this.createIframe({ calLink, queryObject: Cal.getQueryObject(config) });
+    const iframe = this.cal.createIframe({ calLink, calOrigin, queryObject: Cal.getQueryObject(config) });
```

Comment:
Because 'modal' is a method on CalApi now, refer to createIframe through this.cal